libc: add tests for malloc() alignment

martin-aberg opened the pull request:
Adds tests verifying that an application can use memory allocated by
`malloc()`, `calloc()` and `realloc()` to access any type of object in a
portable way.

This behavior is required by the C standard and applications should
be able to rely on it. For example:
```
        double *pi = malloc(sizeof (double));
        if (pi) {
                *pi = 3.1;
        }
```

The tests added by this pull request do a number of allocations of
variable size ~and verifies alignment of the returned pointers~. In
addition, the returned pointers are dereferenced to verify that the
application can access the objects without trapping.

~The `minimal` libc implementation has been updated to pass the new tests.~

tbr-tt committed:
libc/minimal: reallocarray() in terms of realloc()

reallocarray() is defined in terms of realloc(). From OpenBSD manual
pages:
    "Designed for safe allocation of arrays, the reallocarray()
    function is similar to realloc() except it operates on nmemb
    members of size size and checks for integer overflow in the
    calculation nmemb * size."

The return value of sys_heap_realloc() is not compatible with that
of realloc().

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

andyross commented on include/sys/sys_heap.h:
```diff
@@ -142,6 +142,26 @@ void sys_heap_free(struct sys_heap *h, void *mem);
+ * @return Pointer to memory the caller can now use, or NULL
+ */
+void *sys_heap_aligned_realloc(struct sys_heap *heap, void *ptr, size_t align,
+			       size_t bytes);
```
Not opposed to the API, exactly, but this mechanism being introduced seems senselessly complicated.  The heap code already guarantees an 8-byte alignment, which is AFAIK sufficient for all data member alignment requirements on all of our platforms.

Can't you just put some BUILD_ASSERT() statements in and call it a day?  Why are we mucking in the heap code to add support for a feature we already have?

martin-aberg replied:
The test case in the PR fails unless the implementation is changed.

For example, if the top commit (`malloc.c`) in the PR is reverted, then the test reports that `malloc()` gives 4-byte alignment on riscv32, xtensa and sparc:
```
    Assertion failed at ZEPHYR_BASE/tests/lib/mem_alloc/src/main.c:140: test_malloc_align: (aligned_to_all_builtin_types(aptr) is false)
0x8000159c: pointer not suitable for built-in types
```

I made an attempt to define `z_malloc_heap_mem` in terms of `uint64_t` (instead of `char`) and verified the aligment in `zephyr.map`. Unfortunately that did not help and `malloc()` still gave 4-byte aligned pointers at run-time.

This is unexpected if the heap is guaranteeing 8-byte alignment on 32-bit ABI. Could something else be wrong in how `malloc.c` uses the heap code, which can clarify this behavior?

A related topic is the question of what kind of alignment we want to provide in the "minimal" implementation of `malloc()` and `realloc()`. The theoretical answer is "to any type of object". A more practical approach would be to make sure we cover at least `alignof()` of `double`, `uint64_t`, `void *` (but skip `long double` and the `complex.h` types).

martin-aberg replied:
After investigating this a bit more I made the following observation of the run-time behavior (not diving into the code):

If `CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE` is `262151` or less, then `sys_heap_alloc()` sometimes returns 4-byte aligned pointers. But if that config is `262152` (256 KiB) or greater, then it appears to always give 8-byte aligned pointers. (The `mem_alloc` test in this PR uses a 2048 byte heap.)

There is a comment at the top of `heap.h` indicating that a different approach is taken by the implementation depending on heap size (256 KiB or larger, vs less). https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/os/heap.h#L27

andyross replied:
I'm still not following.  Types like doubles on all of our 32 bit platforms currently are perfectly safe to align on 4-byte boundaries.  Is this an argument about runtime behavior or standards compliance or what?

If there's a real concern that a specific platform (which one?!) simply can never allow a 4-byte aligned heap block, then the fix should be to use the "big" heap mode, which is precisely what it's for.

martin-aberg replied:
The 32-bit SPARC architecture specification requires that the CPU implementation trap any 64-bit load/store which is not naturally aligned. This includes access to objects of the C types `double` and `uint64_t`. The 8-byte alignment requirement applies for example to the 32-bit in-tree Zephyr board `qemu_leon3` which is run as part of CI. Issue is described in #30723.

I have followed your advice and removed the heap and malloc changes from the PR. The test case for data access is still in. Thank you for your feedback.

Moving forward, using the "big" heap mode sounds like a really good solution. How can it be enabled?

andyross commented on lib/libc/minimal/source/stdlib/malloc.c:
```diff
@@ -35,7 +35,12 @@ Z_GENERIC_SECTION(POOL_SECTION) static char z_malloc_heap_mem[HEAP_BYTES];
+	 * the returned pointer can be used to access any of the types in
+	 * z_max_align_t.
+	 */
+	ret = sys_heap_aligned_alloc(&z_malloc_heap, Z_MAX_ALIGN, size);
```
As mentioned, this criteria is already met.  I'd really prefer this be a build-time check vs. increasing code size.

andyross reviewed (changes requested):
The libc and test changes seem fine, not loving the seemingly-needless hackery to the heap code.

tbr-tt committed:
tests/mem_alloc: check that malloc() objects can be accessed

Adds tests checking that we can use memory allocated by malloc(),
calloc() and realloc() to access objects of some common types,
including uint64_t and double.

It works by doing a number of allocations of various sizes and
dereferencing the returned pointer. The purpose is to catch cases where
the application would trap if accessing the allocated memory. (The
test does not check alignment against the ABI or alignof().)

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

andyross reviewed (approved):
These all look good.  I'm still not undertsanding whether or not the specific problem you are describing is real or theoretical.

On platforms where all heap blocks must be 8 byte aligned, the correct fix should be to change the logic to enforce the use of the "big chunks" mode always on that platform.  You can see this code right now in lib/os/heap.h:69, in big_heap_chunks().  Currently it just assumes that it should align to pointer size.  You can augment that however you like, for example to include an architecture test (or even go so far as to define an arch-specific kconfig like CONFIG_ARCH_MIN_HEAP_ALIGN or somesuch).

Does that sound like something you'd be comfortable adding here?  Again, I don't really know the full list of platforms you're trying to fix.

tbr-tt committed:
lib/os/heap: introduce option to force big heap mode

This option allows forcing big heap mode. Useful on for getting 8-byte
aligned blocks on 32-bit machines.

Signed-off-by: Martin Åberg <martin.aberg@gaisler.com>

martin-aberg commented:
Thanks @andyross for the guidance. It is much appreciated.

> Does that sound like something you'd be comfortable adding here?

Sure. I have made an attempt which is now in the commit at the top of this PR. It allows to enforce the big heap mode with a bool config. If the config is not enabled, then the default (previous) way of selecting heap mode is used. The config is enabled only on SPARC for now, so nothing else should be affected. Yes the specific problem described is a real problem.

The change to the heap code is:

```diff
 static inline bool big_heap_chunks(size_t chunks)
 {
+       if (IS_ENABLED(CONFIG_SYS_HEAP_ALWAYS_BIG_MODE)) {
+               return 1;
+       }
        return sizeof(void *) > 4U || chunks > 0x7fffU;
 }
```
